### ISSUE TITLE
Investigation usage of SQLite database

### ISSUE THREAD

**Issue by JurgenR** [JurgenR/aioslsk#242]
Currently the Python `shelve` module is used for storing shares and transfers but this ties the cache to the implementation, if a change needs to be made to the `Transfer`, `SharedItem`, `SharedDirectory` objects then some migration code needs to be written.

Using an SQLite database would be cleaner, but comes with its own set of problems. If done right this could also remove the need to store all shared files and transfer in memory.

There's a case to be made to split `cache` into `cache` and `storage`, with flexibility to mix different types.

Assume a collection of 5 million shared files as a target (this seems roughly the max amount of shared files seen in rooms by users). Transfers are a bit hard to quantify.

**Speed**

When connected to the distributed network the client will receive many search requests. Because shared items are currently stored in memory this is very fast, using an SQLite database would slow this down significantly. There are libraries available to be able to make SQL queries asynchronously and querying the local shares is already done as part of a task. The queries don't actually need to be super fast like implemented today, but they also cannot be slow (eg.: 3s+). If a query on millions of records takes 250ms than this could be considered fine.

More worrying is the insertion and deletion of shared items during scanning, insertion of millions of records can be very slow.

**Migration**

If changes need to be made to the database there needs to be some easy way to do migrations.

The changes could include some migration hooks, try finding a library that does versioning.

**Comment by jbaiter** [comment]
> More worrying is the insertion and deletion of shared items during scanning, insertion of millions of records can be very slow.

I don't think this should be much of an issue if SQLite is used with the right PRAGMAs and the updates are done batched.

> The queries don't actually need to be super fast like implemented today, but they also cannot be slow (eg.: 3s+). If a query on millions of records takes 250ms than this could be considered fine.

Without having looked deeper into it, from my experience with it, that sounds absolutely doable with SQLite, if there's not a ton of joining or expensive where-clauses going on.

Would you be interested in a PR that gives this a shot?

**Comment by JurgenR** [comment]
> Sorry, I wasn't being totally clear, my concern wasn't the memory cost, but rather the added runtime latency for obtaining the size during retrieval for every item in the result set: since FileData requires it, it is calculated on-demand in [convert_item_to_file_data](https://github.com/JurgenR/aioslsk/blob/main/src/aioslsk/shares/utils.py#L50), incuring a synchronous stat syscall for every file in the result set.
Adding the file size to the stored data should give a decent performance boost for retrieval of large result sets, irrespective of the implementation :-)

Indeed, understood and agree :)

> Memory usage hasn't been a large issue so far. I haven't checked how much of a performance impact concatenating and compressing during each request would have.

There's always going to be some kind of trade off for this. 

I haven't tested it either but I don't believe the concatenation would be a big cost. For compression memory usage could be reduced by using [zlib.compressobj](https://docs.python.org/3/library/zlib.html#zlib.compressobj). 

> since we only have three share groups (public, buddy, trusted buddy).

I hadn't even thought about the share response being different due to visibility of files to different users

**Comment by mathiascode** [comment]
> In any case, I'll check how large of an impact concatenating and compressing a response containing millions of files has.

Compressing the response is by far the slowest.

I tested with a response about 132 MB in size uncompressed (~3 million shared files, ~45k folders, almost no audio attributes), and compressing it took about 4 seconds, with a final compressed response size of ~28 MB. Changing the zlib compression level from the default (~6) to 4 cut this time down to 2 seconds, with a final compressed response size of ~30 MB.

I checked the compressed response size of the user with the most shares files I know of (about 6 million files, 700k folders), and it's ~150 MB. My 1-2GB estimation for all the responses stored in memory in Nicotine+ was off, that's closer to the truth for the uncompressed responses at worst.

**Comment by jbaiter** [comment]
Progress report: Working on it, got the database schema and retrieval/query stuff done and am currently working my way through the various parts of the code that update the state of `SharedDirectory` and `SharedItem` objects to see how we best go about keeping the database in sync with the state.
